Avoid storing effective language on ElementRareData when it matches the effective document element language

heycam opened the pull request:
#### 8858f1de792a602c09ff087c13c342014f27fa62
<pre>
Avoid storing effective language on ElementRareData when it matches the effective document element language
<a href="https://bugs.webkit.org/show_bug.cgi?id=251657">https://bugs.webkit.org/show_bug.cgi?id=251657</a>
&lt;rdar://problem/104987630&gt;

Reviewed by Ryosuke Niwa.

When an element&apos;s lang attribute is set, we set the effective language on
the ElementRareData of all elements in the subtree. We have an existing
optimization that avoids this if we&apos;re setting it on the document
element. There are some pages, like Wikipedia, that set an explicit
lang on some other element in the body of the page, but which matches
the document element&apos;s lang.

We can avoid the memory overhead of allocating ElementRareData to store
the effective language in this case by using a flag on the element. On
large Wikipedia pages this can save several MB.

If the document element language changes later, we must update the
effective lang state on subtrees that are using this flag. A WeakHashSet
is added to Document to track elements that have an explicit lang
attribute that matches the effective document element language.

Three flags on EventTarget are introduced:

- HasLangAttr and HasXMLLangAttr: records that the element has the
  corresponding lang attribute. This allows us to avoid searching
  for an attribute when updating the effective lang state.

- EffectiveLangKnownToMatchDocument: records that the element has an
  effective lang that matches the effective document element language,
  whether it&apos;s due to an explicit lang attribute or inherited from an
  ancestor. This flag is used in place of
  ElementRareData::m_effectiveLang.

The EffectiveLangKnownToMatchDocument flag is used in place of the &quot;null
effective language means we&apos;ve inherited the effective document
language&quot; state so that disconnected subtrees can also make use of this
optimization. Otherwise, for a case like this:

  let e = document.createElement(&quot;div&quot;);
  e.lang = &quot;en&quot;;  // matching document
  e.append(document.createElement(&quot;div&quot;));

the child element would not know whether to return nullptr or &quot;en&quot; from
Element::effectiveLang() without looking up the tree to see if there is
an ancestor with a langauge attribute.

The EffectiveLangKnownToMatchDocument flag is not
EffectiveLangMatchesDocument, since we don&apos;t set it if the document
element language changes and an existing element starts matching it.
Rather than track all elements with lang attributes to handle such
cases, we leave the effective lang stored on the ElementRareData.

* LayoutTests/fast/css/lang-matching-document-invalidation-expected.txt: Added.
* LayoutTests/fast/css/lang-matching-document-invalidation.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setDocumentElementLanguage):
(WebCore::Document::addElementWithLangAttrMatchingDocument):
(WebCore::Document::removeElementWithLangAttrMatchingDocument):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::~Element):
(WebCore::Element::attributeChanged):
(WebCore::Element::setEffectiveLangInSubtree):
(WebCore::Element::didMoveToNewDocument):
(WebCore::Element::setEffectiveLangFromParent):
(WebCore::Element::setEffectiveLang):
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::hasEffectiveLangState const):
(WebCore::Element::removedFromAncestor):
(WebCore::Element::effectiveLang const):
(WebCore::Element::langFromAttribute const):
(WebCore::Element::langAttrMatchesDocument const):
(WebCore::Element::setLangAttrMatchesDocument):
(WebCore::Element::effectiveLangMatchesDocument const):
(WebCore::Element::setEffectiveLangMatchesDocument):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/EventTarget.h:

Canonical link: <a href="https://commits.webkit.org/259931@main">https://commits.webkit.org/259931@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc89cbca887c67aa66680b92a5ac7ab415924c69

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106426 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15454 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39246 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115613 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175717 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110333 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16942 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6679 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98628 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115278 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112192 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12866 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95845 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40429 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94741 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27489 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8686 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28841 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9217 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5413 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14840 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48387 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10767 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3694 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -265,6 +265,8 @@ Element::~Element()
             map->clearElement();
     }
+
+    document().removeElementWithExplicitLanguageMatchingDocumentElement(*this);
```
This isn't quite right. An element can move from one document to another.
Please do this in removedFromAncestor instead.
Also, we need to insert the element into the map when an element with existing lang attribute gets inserted into document.

heycam replied:
I think it will be simpler not to use the optimization for elements outside the document. Otherwise we really need to deal with the adoption of the elements to get this right, rather than just when it's inserted and removed.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2603,6 +2614,12 @@ Node::InsertedIntoAncestorResult Element::insertedIntoAncestor(InsertionType ins
+        auto langAttr = langFromAttribute();
+        if (!langAttr.isNull() && langAttr != document().documentElementLanguage())
+            setEffectiveLangInSubtree();
+    }
```
If this element is newly connected to a document, then we must also check if it has lang attribute that's identical to document element's language, in which case we need to add it to the set of elements maintained by document to handle a case like this:
```
<html lang="zh">
<body>
<script>
const element = document.createElement('div');
element.lang = "zh";
document.body.appendChild(element);
document.documentElement = 'ja';
</script>
</html>
```

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2664,9 +2681,14 @@ void Element::removedFromAncestor(RemovalType removalType, ContainerNode& oldPar
         if (auto* parent = parentOrShadowHostElement(); langFromAttribute().isNull()
             && !(parent && UNLIKELY(parent->hasRareData()) && !parent->elementRareData()->effectiveLang().isNull()))
             elementRareData()->setEffectiveLang(nullAtom());
+    } else if (!langFromAttribute().isNull()) {
```
I don't think we want this function call to happen whenever an element is removed.
The function is O(k) where k is the number of attributes on this element.
Instead, we should probably set a bit flag whenever we have an element with an explicit lang attribute.

rniwa replied:
Also, we need to do this work only when `removalType.disconnectedFromDocument` is true.

heycam replied:
A bit flag is what I wanted to use, to avoid searching for the lang attribute, but I could not see a free flag on Node/Element.

rniwa replied:
@heycam : There are more bits left in EventTarget.

heycam replied:
I could give a flag like IsEditingText a dual purpose?

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2099,28 +2099,11 @@ void Element::attributeChanged(const QualifiedName& name, const AtomString& oldV
         } else if (name == HTMLNames::langAttr || name.matches(XMLNames::langAttr)) {
             if (document().documentElement() == this)
                 document().setDocumentElementLanguage(newValue);
+            else if (isConnected() && (!UNLIKELY(hasRareData()) || elementRareData()->effectiveLang().isNull()) && langFromAttribute() == document().documentElementLanguage())
```
I don't think we want to have this isConnected check here.
We probably want to keep track of every element that has the matching lang attribute to be in the set.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2603,6 +2614,12 @@ Node::InsertedIntoAncestorResult Element::insertedIntoAncestor(InsertionType ins
 
+    if (!UNLIKELY(hasRareData()) || !elementRareData()->effectiveLang()) {
+        auto langAttr = langFromAttribute();
+        if (!langAttr.isNull() && langAttr != document().documentElementLanguage())
```
This doesn't look right. Why does ancestor of this element changing anything to do with effective lang getting changed? If we're trying to detect when an element is moved from one document to another, then we should be doing that in Element::didMoveToNewDocument instead.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2603,6 +2614,12 @@ Node::InsertedIntoAncestorResult Element::insertedIntoAncestor(InsertionType ins
     }
 
+    if (!UNLIKELY(hasRareData()) || !elementRareData()->effectiveLang()) {
+        auto langAttr = langFromAttribute();
```
Again, this langFromAttribute call will do O(k) traversal where k is the number of attributes on this element so we really want to avoid that by using a bit on EventTarget.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5205,4 +5254,24 @@ void Element::setDisplayContentsChanged(bool changed)
 
+bool Element::langAttrMatchesDocument() const
+{
+    return hasEventTargetFlag(EventTargetFlag::LangAttrMatchesDocument);
```
We might as inline these.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5205,4 +5254,24 @@ void Element::setDisplayContentsChanged(bool changed)
+
+bool Element::effectiveLangMatchesDocument() const
+{
+    return hasEventTargetFlag(EventTargetFlag::EffectiveLangMatchesDocument);
```
Ditto

rniwa commented on Source/WebCore/dom/Element.h:
```diff
@@ -842,6 +844,16 @@ class Element : public ContainerNode {
+    void setEffectiveLang();
+    void setEffectiveLangFromParent();
+
+    bool langAttrMatchesDocument() const;
```
I think we should be explicit in calling out document element since there is also lang set forth by HTTP header. To say "matches the document" leaves this a bit ambiguous.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2498,6 +2499,52 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+
+void Element::setEffectiveLang()
+{
+    auto& lang = langFromAttribute();
```
Hm... this would mean every adoption of an element results in O(k) traversal where k is the number of attributes.

Can we instead remember if this element has an explicit lang attribute set or not using a EventTarget bit flag? Then the combination of that flag plus the flag indicating that the effective lang matches that of the document element should give us logically equivalent sets of flags.

heycam replied:
Unfortunately we'd still need an extra bit, since we'd need to track whether the element has a `lang` and an `xml:lang`. Unless upon removal of one we search for whether the other one is present.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2498,6 +2499,52 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+    }
+}
+
+void Element::setEffectiveLang()
```
I'd call this updateEffectiveLangStates or updateEffectiveLangBitFlags.
We're not actually updating the effective lang. That happens in setEffectiveLangInSubtree.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2141,6 +2123,25 @@ void Element::attributeChanged(const QualifiedName& name, const AtomString& oldV
         cache->deferAttributeChangeIfNeeded(this, name, oldValue, newValue);
 }
 
+void Element::setEffectiveLangInSubtree()
```
How about recursivelyUpdateEffectiveLang?
This function doesn't necessarily update the effective lang for the whole subtree, skipping any sub-subtree with a lang attribute of its own so it's more fitting to me to put the emphasis on recursive nature of the algorithm instead of subtree.

Or maybe updateEffectiveLangOnDescendants.

rniwa reviewed (commented):
I *think* this patch is correct but let me think about it some more.

rniwa commented on LayoutTests/fast/css/lang-matching-document-invalidation.html:
```diff
@@ -0,0 +1,231 @@
+    await new Promise(function(resolve) {
+        otherFrame.onload = () => resolve();
+        otherFrame.srcdoc = `<!DOCTYPE html><html lang=en><body>`;
+        document.body.append(otherFrame);
```
No need to do this. Just insert iframe & start accessing contentDocument. Everything happens synchronously in WebKit/Blink.

rniwa commented on LayoutTests/fast/css/lang-matching-document-invalidation.html:
```diff
@@ -0,0 +1,231 @@
+
+function test(f) {
+    document.documentElement.lang = "en";
+    try {
```
I don't think try & catch is needed with js-test.js.

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1541,6 +1541,11 @@ void Document::setDocumentElementLanguage(const AtomString& language)
 
+    for (auto& element : copyToVector(m_elementsWithLangAttrMatchingDocument)) {
+        if (element)
+            element->setEffectiveLangInSubtree();
```
Don't we need to upset the flag on every element here?

Also, this set is now stale, right? There will be another set of elements that matches the new lang value of the document element.

heycam replied:
The flag is updated and the element removed from the WeakHashSet by `setEffectiveLangInSubtree`.

But you are right that the set will be stale in that there may now be elements in the document that have a `lang=""` that matches the new document element lang, but which are not in the WeakHashSet or have the right bits set. This is OK, since on those elements we still have the effective lang stored on the ElementRareData. We could be clearer about this by calling the flag `EffectiveLangKnownToMatchDocumentElement` or something.

heycam replied:
>The flag is updated and the element removed from the WeakHashSet by setEffectiveLangInSubtree.

Though it would be slightly more efficient to clear the WeakHashSet here, as we'll avoid churning the table as we process each element.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2498,6 +2501,47 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
     }
+
+    if (langAttrKnownToMatchDocumentElement())
+        oldDocument.removeElementWithLangAttrMatchingDocumentElement(*this);
```
I think we wanna clear the flag here. Otherwise, I'd try try to remove this element from new document's set in updateEffectiveLangState, which seems like a wasteful hashmap look up.

heycam replied:
FWIW the WeakHashSet lookup at this point is cheap because we've cleared it and the `find()` call will return early without even hashing the input value. But I can do this.

rniwa replied:
I don't follow. We're just adopting this element to a new document. That new document can have any number of existing elements with lang matching the document element, right?

heycam replied:
Sorry, I overlooked the word "new" in your comment, and you are right that we should be clearing the flag here.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2498,6 +2501,47 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+            ensureElementRareData().setEffectiveLang(parent->elementRareData()->effectiveLang());
+        else if (hasRareData())
+            elementRareData()->setEffectiveLang(nullAtom());
+    }
```
What happens when the parent element is document element here?

heycam replied:
We need to handle a transition between "has non-document-element-matching lang" to "no lang" for such elements. I'll add a new subtest.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -265,6 +265,9 @@ Element::~Element()
             map->clearElement();
     }
+
+    if (langAttrKnownToMatchDocumentElement())
```
By our naming convention, this would probably be renamed to hasLangAttrKnownToMatchDocumentEleement.

heycam committed:
Avoid storing effective language on ElementRareData when it matches the effective document element language
https://bugs.webkit.org/show_bug.cgi?id=251657
<rdar://problem/104987630>

Reviewed by Ryosuke Niwa.

When an element's lang attribute is set, we set the effective language on
the ElementRareData of all elements in the subtree. We have an existing
optimization that avoids this if we're setting it on the document
element. There are some pages, like Wikipedia, that set an explicit
lang on some other element in the body of the page, but which matches
the document element's lang.

We can avoid the memory overhead of allocating ElementRareData to store
the effective language in this case by using a flag on the element. On
large Wikipedia pages this can save several MB.

If the document element language changes later, we must update the
effective lang state on subtrees that are using this flag. A WeakHashSet
is added to Document to track elements that have an explicit lang
attribute that matches the effective document element language.

Three flags on EventTarget are introduced:

- HasLangAttr and HasXMLLangAttr: records that the element has the
  corresponding lang attribute. This allows us to avoid searching
  for an attribute when updating the effective lang state.

- EffectiveLangKnownToMatchDocument: records that the element has an
  effective lang that matches the effective document element language,
  whether it's due to an explicit lang attribute or inherited from an
  ancestor. This flag is used in place of
  ElementRareData::m_effectiveLang.

The EffectiveLangKnownToMatchDocument flag is used in place of the "null
effective language means we've inherited the effective document
language" state so that disconnected subtrees can also make use of this
optimization. Otherwise, for a case like this:

  let e = document.createElement("div");
  e.lang = "en";  // matching document
  e.append(document.createElement("div"));

the child element would not know whether to return nullptr or "en" from
Element::effectiveLang() without looking up the tree to see if there is
an ancestor with a langauge attribute.

The EffectiveLangKnownToMatchDocument flag is not
EffectiveLangMatchesDocument, since we don't set it if the document
element language changes and an existing element starts matching it.
Rather than track all elements with lang attributes to handle such
cases, we leave the effective lang stored on the ElementRareData.

* LayoutTests/fast/css/lang-matching-document-invalidation-expected.txt: Added.
* LayoutTests/fast/css/lang-matching-document-invalidation.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setDocumentElementLanguage):
(WebCore::Document::addElementWithLangAttrMatchingDocument):
(WebCore::Document::removeElementWithLangAttrMatchingDocument):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::~Element):
(WebCore::Element::attributeChanged):
(WebCore::Element::setEffectiveLangInSubtree):
(WebCore::Element::didMoveToNewDocument):
(WebCore::Element::setEffectiveLangFromParent):
(WebCore::Element::setEffectiveLang):
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::hasEffectiveLangState const):
(WebCore::Element::removedFromAncestor):
(WebCore::Element::effectiveLang const):
(WebCore::Element::langFromAttribute const):
(WebCore::Element::langAttrMatchesDocument const):
(WebCore::Element::setLangAttrMatchesDocument):
(WebCore::Element::effectiveLangMatchesDocument const):
(WebCore::Element::setEffectiveLangMatchesDocument):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/EventTarget.h:

Canonical link: https://commits.webkit.org/259931@main
